flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2249,14 +2305,585 @@ class ScalarExprLowering {
     return gen(x);
   }
 
-private:
-  mlir::Location location;
+private:
+  mlir::Location location;
+  Fortran::lower::AbstractConverter &converter;
+  fir::FirOpBuilder &builder;
+  Fortran::lower::StatementContext &stmtCtx;
+  Fortran::lower::SymMap &symMap;
+  bool inInitializer;
+  bool useBoxArg{false}; // expression lowered as argument
+};
+} // namespace
+
+// Helper for changing the semantics in a given context. Preserves the current
+// semantics which is resumed when the "push" goes out of scope.
+#define PushSemantics(PushVal)                                                 \
+  [[maybe_unused]] auto pushSemanticsLocalVariable97201 =                      \
+      Fortran::common::ScopedSet(semant, PushVal);
+
+static bool isAdjustedArrayElementType(mlir::Type t) {
+  return fir::isa_char(t) || fir::isa_derived(t);
+}
+static bool elementTypeWasAdjusted(mlir::Type t) {
+  if (auto ty = t.dyn_cast<fir::ReferenceType>())
+    return isAdjustedArrayElementType(ty.getEleTy());
+  return false;
+}
+static mlir::Type adjustedArrayElementType(mlir::Type t) {
+  return isAdjustedArrayElementType(t) ? fir::ReferenceType::get(t) : t;
+}
+
+/// Extract the subelement type and perform any trivial conversions on `val`,
+/// such as `index -> i32` or `i1 -> !fir.logical`, to align the value's type to
+/// array storage.
+static mlir::Value adjustedArrayElement(mlir::Location loc,
+                                        fir::FirOpBuilder &builder,
+                                        mlir::Value val, mlir::Type arrTy) {
+  auto eleTy = fir::unwrapSequenceType(arrTy);
+  return builder.createConvert(loc, adjustedArrayElementType(eleTy), val);
+}
+
+//===----------------------------------------------------------------------===//
+//
+// Lowering of scalar expressions in an explicit iteration space context.
+//
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// In an explicit iteration space, a scalar expression can be lowered
+/// immediately as the explicit iteration space will have already been
+/// constructed. However, the base array expressions must be handled distinctly
+/// from a "regular" scalar expression. Base arrays are bound to fir.array_load
+/// values. Base arrays on the LHS of an assignment must be properly threaded
+/// using block arguments.
+class ScalarArrayExprLowering {
```

Comment:
It's highly likely this can be fused entirely or via inheritance, I agree. There is some interesting "special sauce" in the older scalar lowering code that may need to be sorted out and cleaned up, which didn't seem appropriate for this work.